### PR TITLE
feat: read_delta and to_delta for some backends

### LINKED ISSUE #6319: feat: support reading and writing delta tables
### Is your feature request related to a problem?

currently, I can use use a combination of the Arrow methods in Ibis with either `deltalake` or `polars` for reading and writing Delta tables. however, it would be more convenient to natively support this in Ibis

### Describe the solution you'd like

```python
import ibis

t = ibis.read_delta("path/to/delta/")

# do some transformation

t.to_delta("path/to/delta/", mode="append")
```

polars recently implemented this and could act as a good reference: https://github.com/pola-rs/polars/pull/7616

### What version of ibis are you running?

n/a but 5.1.0

### What backend(s) are you using, if any?

duckdb usually

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by lostmygithubaccount**
closes #6319

adds `read_delta` for some backends and `to_delta` for all backends (writing locally)

TODO:
- [x] tests
~- [ ] datafusion backend~
~- [ ] spark backend?~ 
- [x] handling the dependency/import for deltalake
- [x] update some docstrings

### CODE REVIEW

**Comment by cpcloud** [comment]
Regarding dependencies, I think we can make `deltalake` its own extra, to avoid requiring it for every backend including if someone won't ever need support for writing deltalake files.

I can push up the changes for that.

**Comment by cpcloud** [comment]
Added a small round trip test.

**Comment by cpcloud** [comment]
One issue with the spark implementation is that the recommended way to do this -- using `delta-spark` -- requires `pyspark` 3.4.0, which has terrible support for numpy 1.24 and pandas 2.

**Review by cpcloud — CHANGES_REQUESTED** [review]

File: ibis/backends/polars/__init__.py [hunk 1]
```diff
@@ -210,6 +210,34 @@ def read_parquet(
         self._tables[table_name] = pl.scan_parquet(path, **kwargs)
         return self.table(table_name)
 
+    def read_delta(
+        self, path: str | Path, table_name: str | None = None, **kwargs: Any
+    ) -> ir.Table:
+        """
+        Register a Delta Lake as a table in the current database.
```
Comment (cpcloud) [inline ibis/backends/polars/__init__.py:217]:
```suggestion
        """Register a Delta Lake as a table in the current database.
```

**Comment by cpcloud** [comment]
Pushed up a fix for the conflict.

File: ibis/backends/duckdb/__init__.py [hunk 2]
```diff
@@ -473,6 +473,45 @@ def read_in_memory(
...
+
+        table_name = table_name or util.gen_name("read_delta")
+
+        from deltalake import DeltaTable
+
+        delta_table = DeltaTable(source_table, **kwargs)
+
+        # TODO: to_batches() doesn't work ("Python Object generator not suitable to be registered as a view")
+        return self.read_in_memory(
+            delta_table.to_pyarrow_dataset().to_table(), table_name=table_name
```
[the remark this answers is not in the document]
Comment (lostmygithubaccount) [reply ibis/backends/duckdb/__init__.py:512]: does this need to be updated? I assumed only table/batch reader

**Comment by lostmygithubaccount** [comment]
I think I'll skip Spark for this PR and can add later if requested? thanks for the help here! I'll look to add datafusion, update some docstrings, and look at adding moretests

File: ibis/backends/duckdb/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/duckdb/__init__.py]: Does to_table read the entire thing into memory? If so, we should remove that method call and let it remain a pyarrow dataset.

[the remark this answers is not in the document]
Comment (lostmygithubaccount) [reply ibis/backends/duckdb/__init__.py]: sorry I missed a link -- `to_table` does read it into memory, but pyarrow dataset as an input isn't noted here: https://github.com/ibis-project/ibis/blob/b23689372fa3d9c30651b69541c8945c1be38152/ibis/backends/duckdb/__init__.py#L447

[the remark this answers is not in the document]
Comment (lostmygithubaccount) [reply ibis/backends/duckdb/__init__.py]: and batches didn't seem to work (per my comment above on line 510)

[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/duckdb/__init__.py]: Right, but try it without the to_table call, it should work. We may just need to update the docs/annotations.

**Review by cpcloud — APPROVED** [review]

**Comment by cpcloud** [comment]
Fixed up a small issue from #6323 (replacing `self._tables` assignment with `self._add_table`).

Set to automerge!

**Pull request merged**